### PR TITLE
BIP 141 and 145: GBT updates for SegWit

### PR DESCRIPTION
**Pull request by luke-jr**
@sipa @CodeShark @jl2012

### CODE REVIEW

File: bip-0009.mediawiki [hunk 1]
```diff
@@ -24,6 +24,7 @@ In addition, BIP 34 made the integer comparison (nVersion >= 2) a consensus rule
 
 Each soft fork deployment is specified by the following per-chain parameters (further elaborated below):
 
+# The '''name''' specifies a very brief description of the soft fork, reasonable for use as an identifier.
```
Comment (btcdrak) [inline bip-0009.mediawiki]: I would suggest the for most of the time, most softforks will role out one consensus BIP so the sensible name would be `bipxxxx` unless you have a case like with the bip68+112+113 softfork, in which case we labelled that `csv`. Therefore I would suggest default of `bipxxxx` unless it's a bundle.

Thinking more, we probably need a table of soft forks somewhere with name, bip(s), starttime, timeout.

Comment (luke-jr) [inline bip-0009.mediawiki]: This is outside the scope of this PR, IMO.

**Comment by sipa** [comment]
The nVersion field of a block never affects the rules applied, so a miner can always be free to set the nVersion to anything.

**Comment by luke-jr** [comment]
@sipa Which part are you referring to? GBT servers may very well want to restrict the nVersion to things they can handle.

**Comment by sipa** [comment]
I had interpreted "the server requires support for" as rules the server demanded you to enforce, rather than rules the server demands you to set bits for. It's a bit confusing that rules and rulesenforced are about consensus rules, and rulesrequires is about nVersion bits to be set. Perhaps replace rulesrequired with versionrequired, with an integer of bits the server requires you to set in nversion?

**Comment by luke-jr** [comment]
I would have grouped "rules" and "rulesrequired" together, since they deal with bits available/allowed and required; vs "rulesenforced" which is the actual consensus rules being used. Maybe "vbits" and "vbitsrequired"?

**Comment by jl2012** [comment]
ACK BIP141

**Comment by sipa** [comment]
@Luke vbits would just be the same as version, I think?

Isn't this sufficient:
* vbrequired: bits in nVersion that the server demands to be set (so an
integer)
* rules: list of strings of deployments whose rules are in effect

?

**Comment by luke-jr** [comment]
@sipa How will clients know what bits are *available* and what they mean in context?

**Comment by sipa** [comment]
@luke-jr What do you mean by "available"?

**Comment by luke-jr** [comment]
@sipa Bits which have a defined meaning.

**Comment by sipa** [comment]
@luke-jr Why does that matter? There are never any consensus rules in BIP9 that prevent you from setting any combination at all of version bits. The server may insist that you set some to vote for, but apart from that, that information is not relevant to the client. In particular, setting bits has no effect on consensus rules that are to be enforced for that block.

**Comment by luke-jr** [comment]
The client needs the information to know which bits it wants to or doesn't want to set.

**Comment by sipa** [comment]
They can find that in the BIP.

If they need to rely on the server for that, the server can claim anything,
including claiming that there is nothing to vote for.

**Comment by luke-jr** [comment]
So put each and every BIP into the GBT client, even before it has activated?

Consider that the server trusted for name:bit mappings could very well be your own local bitcoind.

**Comment by luke-jr** [comment]
Also, *even if* each BIP were put into the GBT client, the client doesn't have the required *blockchain context* to determine if the BIP has activated or is still pending.

**Comment by sipa** [comment]
If the client software doesn't know about a BIP, why does it care?

How does it help that the server says "these arbitrary string names exist",
if the client doesn't know what protocol changes they correspond to? If
they care about the rules, they know the rules anyway.

**Comment by sipa** [comment]
@Luke Yes, saying with rules are active is useful. Saying which exist
isn't. They're just arbitrary strings with no meaning to the client. If
they do know about them, they don't need to be told.

**Comment by luke-jr** [comment]
For example, a miner might want to use --veto-vbits=bip109 to instruct their software to automatically unset the BIP 109 bit, and reject pools that require it. Or --require-vbits=bip109 for the opposite.

**Comment by sipa** [comment]
@luke-jr What if the server doesn't know about a particular BIP yet, but the miner does? If the miner has an opinion about a particular BIP/deployment, they also know what bits to use.

**Comment by luke-jr** [comment]
If the server doesn't support it, it shouldn't be indicating it does regardless.

**Comment by sipa** [comment]
Well by default nVersion will always be setting bits for everything that it supports. The miner is always free to unset bits from it, but I don't think it's the server's job to explain the consensus rules to the miner.

**Comment by luke-jr** [comment]
@sipa @petertodd @gmaxwell @rustyrussell @CodeShark @jl2012

This has been revised per discussion. Please re-check and ACK if acceptable.

**Comment by sipa** [comment]
The changes in BIP22 are just formatting?

**Comment by luke-jr** [comment]
Yes, someone noticed the {{templates}} were not rendering on GitHub.

**Comment by jl2012** [comment]
ACK 141

**Comment by theuni** [comment]
Concept ACK on GBT changes. It's a bit hazy though as to what the server should do if a softfork rule is missing from the request, and there's the possibility of altering the block (potentially in a nasty way, like coinbase-only) to make it compatible. In segwit's case, would the lack of a "segwit" rule in the request cause an error post-activation, or would witness transactions be filtered out?

**Comment by luke-jr** [comment]
> It's a bit hazy though as to what the server should do if a softfork rule is missing from the request, and there's the possibility of altering the block (potentially in a nasty way, like coinbase-only) to make it compatible. In segwit's case, would the lack of a "segwit" rule in the request cause an error post-activation, or would witness transactions be filtered out?

The server has a few options:

* Filter out witness transactions and use rules/force
* Produce the full witness-enabled template, and do *not* use rules/force. The client in this case must refuse to proceed unless it can handle segwit (listed in "rules")
* Throw an error

**Comment by theuni** [comment]
Right, and that ambiguity is kinda nasty without a proposed implementation. I'd go as far as saying that it should _never_ throw an error if there's the possibility of somehow creating a valid block. Maybe another input flag could be added for "fail for degraded blocks".

Regardless, this needs to be laid out in the BIP.

**Comment by luke-jr** [comment]
Which method a server uses in this situation is implementation-specific, not part of the standard. In any case, it is already documented clearly here: https://github.com/bitcoin/bips/pull/365/files#diff-6a9961b118279d111faf86ad6ebbfedbR204

Please feel free to proposed improved language.

**Comment by theuni** [comment]
NACK that. Where preferences can be communicated, they should be, and they should be respected by the server. Otherwise what ends up being "part of the standard" is the default behavior of the leading implementation.

**Comment by luke-jr** [comment]
I don't understand what you're NACKing.

**Comment by theuni** [comment]
I'm NACKing the language that makes the behavior implementation-specific. I'm suggest changes to either:
- Add a ```rules/degraded``` value to the template, so that the miner is aware that the block has been downgraded in some way as a result of a missing input rule. This is different from the ```rules/force``` key, in that the latter does not necessarily mean that the block has been degraded.
or
- Add an optional "error_degraded" key to the request that signals that the server should not try to degrade the block as a result of a missing input rule, but throw an error instead.

Along with the addition of text along the lines of ```Unless signaled (by whichever mechanism above), the server should always attempt to provide the client with a template that will allow the client to create a valid block, degraded only as much as is necessary. Only in the case where no solution from the client would be valid should the server return an error.```

With those changes, it would be clear what the implementation for Core should look like.

**Comment by luke-jr** [comment]
Any item in "rules" the client does not understand implies it must error or (if "rules/force" is allowed) "is degraded".

**Comment by theuni** [comment]
Force (as I read it) doesn't necessarily mean degraded. It could simply mean that a serialization change would break the block, but that transactions haven't been removed. For example some future soft-fork that specifies a required order for transactions.

More importantly though, the text above guarantees backwards-compatibility when possible, as opposed to letting the implementation decide.

**Comment by luke-jr** [comment]
A softfork requiring ordered transactions would necessarily need to a new "rules" entry. When there are no rules that the client doesn't understand, "rules/force" is a no-op.

It's not the business of the spec to force backward compatibility. Implementations that don't want to provide it should be free not to.

File: bip-0009.mediawiki [hunk 2]
```diff
@@ -157,6 +161,45 @@ block, indexed by its parent.
...
+
+Softfork deployment names listed in "rules" or as keys in "vbavailable" may be prefixed by a '!' character.
+Without this prefix, GBT clients may assume the rule will not impact usage of the template as-is; typical examples of this would be when previously valid transactions cease to be valid, such as BIPs 16, 65, 66, 68, 112, and 113.
+If a client does not understand a rule without the prefix, it may use it unmodified for mining.
+On the other hand, when this prefix is used, it indicates a more subtle change to the block structure or generation transaction; examples of this would be BIP 34 (because it modifies coinbase construction) and 141 (since it modifies the txid hashing and adds a commitment to the generation transaction).
+A client that does not understand a rule prefixed by '!' must not attempt to process the template, and must not attempt to use it for mining even unmodified.
+
+Servers should only signal this mutation when they have verified a client behaving this way will not produce invalid blocks.
```
Comment (sipa) [inline bip-0009.mediawiki]: Does this section still match the text before?

Comment (luke-jr) [inline bip-0009.mediawiki]: No, it was left over. Removed.

File: bip-0141.mediawiki [hunk 3]
```diff
@@ -120,9 +120,11 @@ The new rule is ''block cost'' ≤ 4,000,000.
 
 Sigops per block is currently limited to 20,000. We change this restriction as follows:
 
-''Sigop cost'' is defined. The cost of a sigop in traditional script is 4, while the cost of a sigop in witness program is 1.
+Sigops in the current pubkey script, signature script, and P2SH check script are counted at 4 times their previous value.
```
Comment (sipa) [inline bip-0141.mediawiki]: Any reason for removing the sigop cost terminology? I thought it was useful to distinguish from the original sigops definition.

**Comment by sipa** [comment]
@theuni Have your concerns been addressed?

File: bip-0141.mediawiki (hunk 3 above)
Comment (luke-jr) [inline bip-0141.mediawiki]: I don't see how the original sigops definition is relevant post-segwit. ie, what's the point?

Comment (sipa) [inline bip-0141.mediawiki]: Avoiding confusion :)

Comment (luke-jr) [inline bip-0141.mediawiki]: I personally find it more confusing to have two terms for essentially the same thing of different versions... should GBT be changed to remove sigops/sigoplimit and add sigopcost/sigopcostlimit as well?

**Comment by sipa** [comment]
ACK
